src/webgpu/shader/execution/expression/call/builtin/textureDimensions.spec.ts

```diff
@@ -0,0 +1,514 @@
+      const w = Math.max(bw, kMinLen) * kMultipleC;
+      const h = Math.max(bh, kMinLen) * kMultipleB;
+      return { size: [w, h, 4], expected: [w >>> mip, h >>> mip] };
+    }
+    case '3d': {
+      const w = Math.max(bw, kMinLen) * kMultipleA;
+      const h = Math.max(bh, kMinLen) * kMultipleB;
+      const d = kMinLen * kMultipleC;
+      return {
+        size: [w, h, d],
+        expected: [w >>> mip, h >>> mip, d >>> mip],
+      };
+    }
+    case 'cube': {
+      const l = Math.max(bw, kMinLen) * Math.max(bh, kMinLen) * kMultipleB;
+      return {
+        size: [l, l, kNumCubeFaces],
+        expected: [l >>> mip, l >>> mip],
+      };
+    }
+    case 'cube-array': {
+      const l = Math.max(bw, kMinLen) * Math.max(bh, kMinLen) * kMultipleC;
+      return {
+        size: [l, l, kNumCubeFaces * 3],
+        expected: [l >>> mip, l >>> mip],
+      };
+    }
+  }
+}
+
+/**
+ * Builds a shader module with the texture view bound to the WGSL texture with the given WGSL type,
+ * which calls textureDimensions(), assigning the result to an output buffer.
+ * This shader is executed with a compute shader, and the output buffer is compared to
+ * `values.expected`.
+ */
+function run(
+  t: GPUTest,
+  view: GPUTextureView,
+  textureType: string,
+  levelArg: number | undefined,
+  values: TestValues
+) {
+  const outputType = values.expected.length > 1 ? `vec${values.expected.length}u` : 'u32';
+  const wgsl = `
+@group(0) @binding(0) var texture : ${textureType};
+@group(0) @binding(1) var<storage, read_write> output : ${outputType};
+
+@compute @workgroup_size(1)
+fn main() {
+output = ${
+    levelArg !== undefined
+      ? `textureDimensions(texture, ${levelArg})`
+      : 'textureDimensions(texture)'
+  };
+}
+`;
+  const module = t.device.createShaderModule({
+    code: wgsl,
+  });
+  const pipeline = t.device.createComputePipeline({
+    compute: { module },
+    layout: 'auto',
+  });
+  const outputBuffer = t.device.createBuffer({
+    size: 32,
+    usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.STORAGE,
+  });
+  const bindgroup = t.device.createBindGroup({
+    layout: pipeline.getBindGroupLayout(0),
+    entries: [
+      { binding: 0, resource: view },
+      { binding: 1, resource: { buffer: outputBuffer } },
+    ],
+  });
+  const encoder = t.device.createCommandEncoder();
+  const pass = encoder.beginComputePass();
+  pass.setPipeline(pipeline);
+  pass.setBindGroup(0, bindgroup);
+  pass.dispatchWorkgroups(1);
+  pass.end();
+  t.device.queue.submit([encoder.finish()]);
+
+  t.expectGPUBufferValuesEqual(outputBuffer, new Uint32Array(values.expected));
+}
+
+/** @returns true if the GPUTextureViewDimension is valid for a storage texture */
+function dimensionsValidForStorage(dimensions: GPUTextureViewDimension) {
+  switch (dimensions) {
+    case '1d':
+    case '2d':
+    case '2d-array':
+    case '3d':
+      return true;
+    default:
+      return false;
+  }
+}
+
+g.test('sampled_and_multisampled')
```

Comment:
I'm seeing `EXCEPTION: Unexpected validation error occurred: The size ([Extent3D width:16, height:24, depthOrArrayLayers:1]) of the texture is not a multiple of the block width (5) and height (4) of the texture format (TextureFormat::ASTC5x4Unorm).`
on a lot of the weirdly-shaped ASTC formats. Sounds like a test bug.

Reply:
My bad. This got broken during code review. I had tested this across multiple GPUs originally, but apparently only tested on a GPU without ASTC support when addressing code review comments. Fixed in https://github.com/gpuweb/cts/pull/3413.